server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java

```diff
@@ -663,6 +663,27 @@ public static String findV2Template(Metadata metadata, String indexName, @Nullab
             return null;
         }
 
+        // this is complex but if the index is not hidden in the create request but is hidden as the result of template application,
+        // then we need to exclude global templates
+        if (isHidden == null) {
+            final Optional<Map.Entry<IndexTemplateV2, String>> templateWithHiddenSetting = matchedTemplates.entrySet().stream()
+                .filter(entry -> IndexMetadata.INDEX_HIDDEN_SETTING.exists(resolveSettings(metadata, entry.getValue()))).findFirst();
```

Comment:
So this brings a question:

- Should we throw an error (the error below for validation) on index creation for a template that won't actually "win" for the creation?

If the answer is yes, then I think rather than returning only the first one, it would be good to return all templates (in the error message I mean) that exhibit this behavior.

If the answer is no, then we can move this check to be below the candidate sorting and only check the `winner` template.

What do you think?

Reply:
Great points. Thanks for bringing this up. 

I saw this validation as "filter the invalid templates _before_ we find the winner template" (ie. the global templates should not even be considered when finding the winning template). 

The other way around, as I understand it, applying this validation on the winner template would be a no-op (since we don't support inheritance anymore, so settings from the global templates cannot be applied) and would look something along the lines of: 

1. the winning template is a global template - no index.hiddne property is allowed on the global templates - so just return it
2. the winning template is not a global one but it has an `index.hidden` setting set to true in the _resolved_ settings, but the other templates have a lower priority so just return it 
3. the winning template is not a global one and doesn't have `index.hidden` set, so just return it

I do agree that having a lower priority (an explicit template) influence and remove a higher priority template (the global one - as illustrated in this test https://github.com/elastic/elasticsearch/pull/55015/files#diff-4e391d69e5a1fe53728ba9530499a6e8R552-R554) from the list of candidates might be (is) counter-intuitive so I think we either go along with this approach or, given the user has the option to override a global template using the priority, we could just issue a warning that a gloable template is used as the matching one but that it is shadowing  explicit templates that try to configure the index as hidden (I do however think this would go unnoticed in the logs)

Reply:
Ok, we decided to validate only the winner in this particular case. Generally (together with https://github.com/elastic/elasticsearch/pull/55010) we will be failing if the `index.hidden` setting is present in the resolved settings of a global template (regardless of the setting value). 

With regards to this particular validation, if the winner is a global template we'll do a sanity check to make sure it doesn't include the `index.hidden` setting, in which case we'll fail, but otherwise we'll return it as the winner template.